### PR TITLE
PROPOSAL: version numbering scheme using SemVer 2.0 

### PR DESCRIPTION
**Pull request by seanknox**
# Background
---
A constant pain point for acs-engine users and internal teams at Microsoft using acs-engine has been lack of clearly defined versions that can be tracked for stability and feature completeness. There is no process of taking an arbitrary VCS commit hash and assigning it a version number that is both meaningful for humans and machines. Using a version numbering scheme is helpful for answering questions like:

- Am I using the latest version of the project?
- Are there any bug fixes for version 1.5 available?
- Which version of a third-party driver (such as CNI) is compatible with acs-engine?

The aim of this proposal is to establish an official version numbering system blessed by the acs-engine core team which can then be used to drive a release process that is stable and encourages product velocity.

## Proposal
This proposal describes a minimal release procedure of tagging the acs-engine repository.

## Release process

This proposal recommends that acs-engine adopt the [semantic versioning 2.0 standard](http://semver.org/spec/v2.0.0.html) for our version numbering scheme.

Source code is released by tagging (eg. `git tag`) the VCS repository with a string representing a SemVer compatible version number for that release.

This proposal is not restricted to git, any VCS that has the facility to assign a tag-like entity to a revision is supported.

## Tag format

The format of the VCS tag is as follows:

```
v<SemVer>
```

That is, the character v, U+0075, followed directly by a string which is compliant with the [Semantic Versioning 2.0 standard](http://semver.org/spec/v2.0.0.html).

When inspecting a repository, tags which do not fit the format described above must be ignored for the purpose of determining the available release versions.

SemVer requires that a once a version number has been assigned, it must not change, thus a tag, once assigned must not be reused.

## Rationale

In the spirit of doing less and enabling more, this proposal establishes the minimum required for humans and tools to identify released versions by inspecting the source code repository directly. It is informed by the broad support for semantic versioning across other OSS contemporaries like node.js (npm), rust (cargo), javascript (bower), and ruby (rubygems), thereby allowing acs-engine consumers and programmers to benefit from the experiences of these other communities' dependency management ecosystems.

### Who benefits from adopting this proposal ?

This proposal will immediately benefit the downstream consumers of acs-engine. For example:
* Version information encourages acs-engine consumers—both internally at Microsoft and external groups like customers—to build other tools and APIs on top of acs-engine releases. For example, a command that will let users upgrade between minor versions, or update to the latest patch released rather than just the latest HEAD of the repository.
* Many customers rely on released versions of software for their packaging policies. They don't want to pull random git hashes into their archives, they want to pull released versions of the code and have release numbers that give them a sense of how compatible new versions are with previous version. For example, Ubuntu has a policy that they only accept patch releases into their LTS distribution; no major version changes, no minor version changes that include new features, only bug fixes.
* Automated systems like CI can leverage version information to build artifacts, run smoke tests, and catch regressions.
* Version numbering allows us to show version information to users which helps them provide better feedback and issues via GitHub and other channels.

### Why SemVer ?

Applying an opaque release tag is not sufficient, the tag has to contain enough semantic meaning for humans and tools to compare two version numbers and infer the degree of compatibility, or incompatibility between them. This is the goal of semantic versioning.

To cut to the chase, SemVer is not a magic bullet, it cannot force developers to not do the wrong thing, only incentivise them to do the right thing. This property would hold true no matter what version numbering methodology was proposed, SemVer or something of our own concoction.

There is a lot to gain from working from a position of assuming ACS contributors want to do the right thing, not engineer a straight jacket process which prevents them from doing the wrong thing. The ubiquity of gofmt'd code, in spite of the fact the compiler allows a much looser syntax, is evidence of this.

Adherence to a commonly accepted ideal of what constitutes a major, minor and patch release is informed by the same social pressures that drive Go programmers to gofmt their code.

### Why not allow the v prefix to be optional ?

The recommendation to include the `v` prefix is for compatibility with the three largest Go projects, Docker, Kubernetes, and CoreOS, who have already adopted this form.

Permitting the `v` prefix to be optional would mean some authors adopt it, and others do not, which is a poor position for a standard. In the spirit of gofmt, mandating the `v` prefix across the board means there is exactly one tag form for implementations to parse, and outweighs the personal choice of an optional prefix.

## Implementation
- A summary of this proposal, along with examples and link to this proposal will be added to relevant internal ACS engineering and product documentation. (_Sean: Anthony/Seth/Cole I look to you to help inform this via this PR and I'll update._)
- Meetings with engineering team(s) as necessary, based on feedback from Engineering Managers.

## Out of scope

The following items are out of scope of this proposal:
- The actual release process implementation using this version numbering—that can be discussed after acceptance of this proposal.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/642)

### CODE REVIEW

**Comment by seanknox** [comment]
For your consideration: @colemickens @anhowe @sgoings @dmitsh. Please tag all other relevant folks.

**Comment by seanknox** [comment]
@jdumars @lachie83 Would love your thoughts as well.

**Comment by colemickens** [comment]
I like the SemVer and the tag syntax.

Do you have examples of projects using the `Upcoming - Minor` and `Upcoming - Major` designations? Where are those specified - via "milestones" in GitHub or in proposal-style release documents? These designations seem a bit confusing to me... why do this instead of just having X.Y.Z milestones with certain Issues or PRs accruing to those defined milestones? For ex, what does an Issue being tag "Upcoming-Minor" mean when looking at it six months after it was filed/fixed/etc?

**Comment by seanknox** [comment]
> Do you have examples of projects using the Upcoming - Minor and Upcoming - Major designations?

I unashamedly stole this from the Helm project's release process, which I think shares some characteristics with this project.

>  Where are those specified - via "milestones" in GitHub or in proposal-style release documents? 

I was thinking GitHub to encourage community participation. What do you think?

> These designations seem a bit confusing to me... why do this instead of just having X.Y.Z milestones with certain Issues or PRs accruing to those defined milestones? For ex, what does an Issue being tag "Upcoming-Minor" mean when looking at it six months after it was filed/fixed/etc?

That's fair. We could try just X.Y.Z milestones with selected issues/PRs to start. IIRC, Helm worked on a two week release cadence, so issue cruft wasn't a big factor. @sgoings can probably talk more about that or @slack could talk more about that.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+## Issues
+Issues are used as the primary method for tracking anything to do with the acs-engine project.
+
+### Issue Lifecycle
+The issue lifecycle is mainly driven by the core maintainers, but is good information for those
+contributing to acs-engine. All issue types follow the same general lifecycle. Differences are noted below.
+1. Issue creation
+2. Triage
+    - The maintainer in charge of triaging will apply the proper labels for the issue. This
+    includes labels for priority, type, and metadata (such as "orchestrator/k8s"). The only issue
```
Comment (anhowe) [inline CONTRIBUTING.md]: Currently our priority has been the following:
1. P0 - means we should fix in the next few weeks
2. P1 - means we should fix in the next few months
3. P2 - let's leave it open to see what interest this gains, but if it sits dormant we close after 3-6 months.

**Comment by anhowe** [comment]
+1 to keeping with the X.Y.Z milestones.  This looks great, I've just added a comment to how we are currently prioritizing.

File: CONTRIBUTING.md (hunk 1 above)
Comment (seanknox) [inline CONTRIBUTING.md]: Will change to reflect the current priority tree.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+and/or it breaks compatibility, then it should be added to the `Upcoming - Major` milestone.
+An issue that we are not sure we will be doing will not be added to any milestone.
+
+A milestone (and hence release) is considered done when all outstanding issues/PRs have been closed or moved to another milestone.
+
+## Issues
+Issues are used as the primary method for tracking anything to do with the acs-engine project.
+
+### Issue Lifecycle
+The issue lifecycle is mainly driven by the core maintainers, but is good information for those
```
Comment (lachie83) [inline CONTRIBUTING.md]: It's probably worth establishing a list of core maintainers and making that available in the repo

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+`2.2.1`, `2.3.0`, `Upcoming - Minor`, or `Upcoming - Major`. If an issue pertains to a
+specific upcoming bug or minor release, it would go into `2.2.1` or `2.3.0`. If the issue/PR
+does not have a specific milestone yet, but it is likely that it will land in a `2.X` release,
+it should go into `Upcoming - Minor`. If the issue/PR is a large functionality add or change
+and/or it breaks compatibility, then it should be added to the `Upcoming - Major` milestone.
+An issue that we are not sure we will be doing will not be added to any milestone.
+
+A milestone (and hence release) is considered done when all outstanding issues/PRs have been closed or moved to another milestone.
+
+## Issues
```
Comment (lachie83) [inline CONTRIBUTING.md]: How are features/featurerequests going to be handled?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+signed the CLA can be accepted into the repository.
+
+
+## Support Channels
+
+Whether you are a user or contributor, official support channels include:
+
+GitHub issues:
+- ACS-Engine: https://github.com/Azure/acs-engine/issues - file issues and PRs related to ACS-Engine
+- ACS: https://github.com/Azure/acs/issues - file issues and PRs related to Azure Container Service
```
Comment (lachie83) [inline CONTRIBUTING.md]: Is it worth establishing a weekly meeting cadence for the community to connect with?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+We'd love to accept your patches! Before we can take them, we have to jump a
+couple of legal hurdles.
+
+The [Microsoft CLA](https://cla.microsoft.com/) must be signed by all contributors. Please fill out either the individual or corporate Contributor License Agreement (CLA). Once you are CLA'ed, we'll be able to accept your pull requests. Once you are CLA'ed, we'll be able to accept your pull requests.
+
+***NOTE***: Only original source code from you and other people that have
+signed the CLA can be accepted into the repository.
+
+
+## Support Channels
```
Comment (lachie83) [inline CONTRIBUTING.md]: Do we need to be more clear about what support means?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    - "Feature" and "Bug" issues should be connected to the PR that resolves it.
+    - Whoever is working on a "Feature" or "Bug" issue (whether a maintainer or someone from
+    the community), should either assign the issue to them self or make a comment in the issue
+    saying that they are taking it.
+    - "Proposal" and "Question" issues should stay open until resolved or if they have not been
+    active for more than 30 days. This will help keep the issue queue to a manageable size and
+    reduce noise. Should the issue need to stay open, the `keep open` label can be added.
+4. Issue closure
+
+## How to Contribute a Patch
```
Comment (lachie83) [inline CONTRIBUTING.md]: Any any requirements on the code having tests?

File: CONTRIBUTING.md (hunk 5 above)
Comment (rgardler) [inline CONTRIBUTING.md]: Yes. As a Microsoft project some customers have an EE ration of "support". This section should essentially say "This is an open source project and as such no formal support is available. However, like all god open source projects we do offer "best effort" support through github issues."

**Comment by rgardler** [comment]
Other than my minor comments in review I'm +1 on this. Very much needed.

File: CONTRIBUTING.md (hunk 6 above)
Comment (seanknox) [inline CONTRIBUTING.md]: I think it'll be good to call this out in release process, which I'll start discussion on after this PR.

File: CONTRIBUTING.md (hunk 4 above)
Comment (seanknox) [inline CONTRIBUTING.md]: That's a great suggestion. Let's parking lot for now until we get a big further in defining releases.

File: CONTRIBUTING.md (hunk 3 above)
Comment (seanknox) [inline CONTRIBUTING.md]: Will call this out in the release proposal.

File: CONTRIBUTING.md (hunk 2 above)
Comment (seanknox) [inline CONTRIBUTING.md]: Good suggestion.

**Comment by seanknox** [comment]
@rgardler @anhowe @colemickens @lachie83 updated per your feedback.

**Comment by seanknox** [comment]
Also added a maintainers list. Let me know if additional folks should be added. Can always do it later, too.

**Comment by dmitsh** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
@acs-bot test this please

**Pull request merged**